<launch>
    <arg name="rviz" default="true" />
    <arg name="scanPeriod" default="0.1" />
    <!-- Set to true if Kitti odometry dataset is used -->
    <arg name="pointUndistorted" default="false" />
    <!-- Set to HDL-64E if Kitti odometry dataset is used -->
    <arg name="lidar" default="VLP-16" />
    <arg name="publishMetrics" default="true" />
    <arg name="listenOdometry" default="true" />
    <!-- Set to false when not visualizing the output -->
    <arg name="publishFullPointCloud" default="true" />

    <node pkg="loam_velodyne" type="laserOdometry"
          name="laserOdometry" output="screen" respawn="true">
        <param name="scanPeriod" value="$(arg scanPeriod)" />
        <!-- <param name="ioRatio" value="2" /> -->
        <!-- <param name="maxIterations" value="25" /> -->
        <!-- <param name="deltaTAbort" value="0.1" /> -->
        <!-- <param name="deltaRAbort" value="0.1" /> -->
        <!-- <param name="residualScale" value="0.05" /> -->
        <!-- <param name="eigenThresholdTrans" value="10.0" /> -->
        <!-- <param name="eigenThresholdRot" value="10.0" /> -->
        <!-- <param name="weightDecayCorner" value="1.8" /> -->
        <!-- <param name="weightThresholdCorner" value="0.1" /> -->
        <!-- <param name="sqDistThresholdCorner" value="25.0" /> -->
        <!-- <param name="weightDecaySurface" value="1.8" /> -->
        <!-- <param name="weightThresholdSurface" value="0.1" /> -->
        <!-- <param name="sqDistThresholdSurface" value="25.0" /> -->
        <param name="pointUndistorted" value="$(arg pointUndistorted)" />
        <param name="publishFullPointCloud"
               value="$(arg publishFullPointCloud)" />
        <param name="publishMetrics" value="$(arg publishMetrics)" />
    </node>

    <node pkg="loam_velodyne" type="laserMapping"
          name="laserMapping" output="screen">
        <param name="scanPeriod" value="$(arg scanPeriod)" />
        <!-- <param name="maxIterations" value="10" /> -->
        <!-- <param name="deltaTAbort" value="0.05" /> -->
        <!-- <param name="deltaRAbort" value="0.05" /> -->
        <!-- <param name="cornerFilterSize" value="0.2" /> -->
        <!-- <param name="surfaceFilterSize" value="0.4" /> -->
        <!-- <param name="mapFilterSize" value="0.0" /> -->
        <!-- <param name="residualScale" value="1.0" /> -->
        <!-- <param name="eigenThresholdTrans" value="100.0" /> -->
        <!-- <param name="eigenThresholdRot" value="100.0" /> -->
        <!-- <param name="sqDistThresholdCorner" value="1.0" /> -->
        <!-- <param name="weightDecayCorner" value="0.9" /> -->
        <!-- <param name="weightThresholdCorner" value="0.1" /> -->
        <!-- <param name="sqDistThresholdSurface" value="1.0" /> -->
        <!-- <param name="weightDecaySurface" value="0.9" /> -->
        <!-- <param name="weightThresholdSurface" value="0.1" /> -->
        <param name="publishFullPointCloud"
               value="$(arg publishFullPointCloud)" />
        <param name="publishMetrics" value="$(arg publishMetrics)" />
    </node>

    <node pkg="loam_velodyne" type="transformMaintenance"
          name="transformMaintenance" output="screen">
        <param name="publishMetrics" value="$(arg publishMetrics)" />
    </node>

    <group if="$(arg listenOdometry)">
        <node pkg="loam_velodyne" type="odometryListener"
              name="odometryListener" output="screen" />
    </group>

    <group if="$(arg publishMetrics)">
        <node pkg="loam_velodyne" type="metricsListener"
              name="metricsListener" output="screen" />
    </group>

    <group if="$(arg rviz)">
        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz"
              args="-d $(find loam_velodyne)/rviz_cfg/loam_velodyne.rviz" />
    </group>
</launch>
